devicedata/generic.py

```diff
@@ -1,20 +1,17 @@
+from django.conf import settings
 from django.utils import timezone
-from devicedata.models import ProvidedData
-from devicedata.providers.opsi import OpsiProvider
-from devicedata.providers.puppet import PuppetProvider
+from django.utils.module_loading import import_string
 
-data_providers = {
-    "opsi": OpsiProvider,
-    "puppet": PuppetProvider
-}
+from devicedata.models import ProvidedData
 
 
 def _get_provider(device):
-    if device.data_provider is not None and device.data_provider in data_providers.keys():
-        return data_providers[device.data_provider]()
+    if device.data_provider in settings.DATA_PROVIDERS:
+        provider_path = settings.DATA_PROVIDERS[device.data_provider]
+        return import_string(provider_path)()
     else:
-        for provider in data_providers.values():
-            provider_instance = provider()
+        for provider_path in settings.DATA_PROVIDERS.values():
+            provider_instance = import_string(provider_path)()
```

Comment:
I've never used the import_string method before. Is there a disadvantage to using it like this in a for loop instead of importing the providers once when the Django app starts?

Reply:
This is the same mechanism that is also used for django settings like `INSTALLED_APPS`, `MIDDLEWARE`, or `AUTHENTICATION_BACKENDS`. I chose it mostly for consistency with those.

I had some issues with exceptions when starting the testserver because `PUPPETDB_SETTINGS` and `OPSI_SETTINGS` are used on module load, but they are not defined in `base_settings.py`. Another fix for that particular issue would be to set them to empty dicts in `base_settings.py`. Or we could change the code so that the settings are only used once the providers are initialized, not on module load. I chose to stop loading those module if they are not required in the first place. Maybe we should even do all three.